src/pyqasm/algorithms/solovay_kitaev/basic_approximation.py

```diff
@@ -0,0 +1,102 @@
+"""
+Definition of the basic approximation algorithm for the Solovay-Kitaev theorem.
+"""
+
+import numpy as np
+
+from pyqasm.algorithms.solovay_kitaev.utils import TU2Matrix, get_tu2matrix_for_basic_approximation
+from pyqasm.elements import BasisSet
+
+
+def rescursive_traversal(
+    target_matrix, approximated_matrix, target_gate_set_list, current_depth, params
+):
+    """Recursively traverse the tree to find the best approximation of the target matrix.
+    Args:
+        target_matrix (np.ndarray): The target matrix to approximate.
+        approximated_matrix (TU2Matrix): The approximated matrix.
+        target_gate_set_list (list): The list of target gates to approximate.
+        current_depth (int): The current depth of the tree.
+        params (tuple): The parameters for the approximation.
+
+    Returns:
+        float: The closest difference between the target and approximated matrix.
+        TU2Matrix: The closest approximated matrix.
+        TU2Matrix: The best approx
+
+    """
+    accuracy, max_tree_depth, closest_diff, closest_gate, best_gate = params
+
+    if current_depth >= max_tree_depth:
+        return closest_diff, closest_gate, best_gate
+
+    if best_gate:
+        return closest_diff, closest_gate, best_gate
+
+    for gate in target_gate_set_list:
+        if not approximated_matrix.can_multiple(gate):
+            continue
+        approximated_matrix_copy = approximated_matrix.copy()
+        approximated_matrix = approximated_matrix * gate
+
+        diff = approximated_matrix.distance(target_matrix)
+        if diff < accuracy:
+            best_gate = approximated_matrix.copy()
+            return closest_diff, closest_gate, best_gate
+
+        # Update the closest gate if the current one is closer
+        if diff < closest_diff:
+            closest_diff = diff
+            closest_gate = approximated_matrix.copy()
+
+        closest_diff, closest_gate, best_gate = rescursive_traversal(
+            target_matrix,
+            approximated_matrix.copy(),
+            target_gate_set_list,
+            current_depth + 1,
+            (accuracy, max_tree_depth, closest_diff, closest_gate, best_gate),
+        )
+        approximated_matrix = approximated_matrix_copy.copy()
+
+    return closest_diff, closest_gate, best_gate
```

Comment:
Should we not return only a single gate i.e. the `best_gate`. I agree that the accuracy might not go below the target accuracy, in which case `closest_gate` becomes the `best_gate`